File: 236-lowest-common-ancestor-of-a-binary-tree/236-lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
      if root:
        if root == p or root == q:
          return root
        lh = self.lowestCommonAncestor(root.left, p, q)
        rh = self.lowestCommonAncestor(root.right, p, q)
        if lh and rh:
          return root
        if(lh):
          return lh
        return rh
      return None
      
        

File: 236-lowest-common-ancestor-of-a-binary-tree/236-lowest-common-ancestor-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* ans;
    bool dfs(TreeNode* node, TreeNode* p, TreeNode* q){
      if(node){
        int l = dfs(node->left, p, q)?1:0;
        int r = dfs(node->right, p, q)?1:0;
        int m = (node == p or node == q)?1:0;
        if(m+l+r >= 2)
          this->ans = node;
        return m+l+r>0;
      }
      return false;
    }
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
      dfs(root, p, q);
      return ans;
    }
};